core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala

```diff
@@ -530,7 +530,7 @@ private[spark] class TaskSetManager(
 
   private def maybeFinishTaskSet() {
     if (isZombie && runningTasks == 0) {
-      sched.taskSetFinished(this)
+      sched.taskSetFinished(this, tasksSuccessful == numTasks)
```

Comment:
actually, I don't quite understand why `tasksSuccessful == numTasks` means an active task set manager is finished. How do we know this is an active task set manager?

Reply:
I think I made a mistake here. Previously I was thinking a zombie TSM must have at least one failed tasks, so tasksSuccessful can't be equal to numTasks. But now I realize that `markPartitionCompleted` could increase `tasksSuccessful` even if it's an already failed task for that zombie TSM. So, it still be possible for a zombie TSM to achieve  `tasksSuccessful == numTasks`. @cloud-fan

Reply:
Up to my knowledge, we can know a stage is finished in `DAGScheduler.markStageAsFinished`. Maybe we can clean up `stageIdToFinishedPartitions` there?

Reply:
Clean up `stageIdToFinishedPartitions` in `DAGSchedule.markStageAsFinished` would be definitely work, though it would add an new place to race for `TaskSchedulerImpl` lock. 

Another way is to distinguish TSM active or zombie  in `markPartitionCompleted()` where an active TSM won't be zombie at the time and clean up `stageIdToFinishedPartitions` right there. Which one do you prefer ? @cloud-fan

Reply:
How do we know a TSM is not zombie in `markPartitionCompleted`?

Reply:
Oh, sorry, it should be `handleSuccessfulTask()` rather than `markPartitionCompleted()`:

https://github.com/apache/spark/blob/bc7592ba1186001127ecfae327ac22a0727b8bab/core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala#L781-L783

where we firstly mark an active TSM into a zombie one.

Reply:
`handleSuccessfulTask` SGTM